README.org
* Atreus Keyboard

The Atreus is a mechanical keyboard designed primarily to match the
shape of human hands and to be as portable as possible. The case
measures 26x12cm and lacks even a number row, relying heavily upon the
=fn= key. There is a circuit board for this design, but it's also
possible to [[http://wiki.geekhack.org/index.php?title=Hard-Wiring_How-To][manually wire the matrix]].

I've seen a number of existing DIY 40% keyboard designs, but most of
them stagger the rows, which I find very annoying now that I've gotten
used to the columnar layout of the [[https://geekhack.org/index.php?topic=22780.0][Ergodox]]. In addition, many of the
designs I've seen waste a lot of room on the space bar, failing to
take into account the fact that the thumb is the strongest and most
versatile of the fingers. This design avoids both these problems while
taking a more couch-friendly single-piece approach.

[[./atreus.jpg]]

See [[./changelog.md][the changelog]] for the various revisions made to the design since
its initial release.

** Kits

You can buy [[http://atreus.technomancy.us][Atreus kits]] that have all the parts you need along with
detailed assembly instructions from http://atreus.technomancy.us. If
you'd rather round up all the parts yourself, that's possible too
since the design is completely open source; see the bill of materials
below.

** Layout

Only a handful of punctuation marks (and no digits) are available
unshifted, and all the modifiers are on the bottom row:

 :    q     w     e     r     t       ||       y     u     i     o    p
 :    a     s     d     f     g       ||       h     j     k     l    ;
 :    z     x     c     v     b       ||       n     m     ,     .    /
 :   esc   tab  super shift bksp ctrl || alt space  fn     -     '  enter

The numbers and most of the punctuation are on the fn layer with a
numpad-style arrangement under the right hand:

 :    !    @     up     {    }        ||     pgup    7     8     9    *
 :    #  left   down  right  $        ||     pgdn    4     5     6    +
 :    [    ]      (     )    &        ||       `     1     2     3    \
 :   L2  insert super shift bksp ctrl || alt space   fn    .     0    =

The =L2= key switches it to the function layer, and tapping =L0= here
brings it back to the first layer.

 :  insert home    ↑    end  pgup     ||       ↑     F7    F8    F9   F10
 :    del   ←      ↓     →   pgdn     ||       ↓     F4    F5    F6   F11
 :                           reset    ||             F1    F2    F3   F12
 :              super shift bksp ctrl || alt space   L0             

The [[https://atreus.technomancy.us/tmk][firmware]] includes a number of other options, including
colemak, dvorak, and "software dvorak" which sends keycodes assuming
the OS will perform the translation into dvorak. Adding new layouts or
changing existing ones is easy.

** Parts

*** Switches

This layout has five modifiers and 37 non-modifiers.

I strongly prefer the feel and sound of tactile [[https://deskthority.net/wiki/Matias_switch#Click][Matias Clicky]]
switches for typing. However, I like having [[https://deskthority.net/wiki/Matias_switch#Linear][Matias Linear
switches]] switches on the modifier keys (ctrl, alt, super, shift, and
fn) because the tactile effect has no benefit for keys that are held
down, and giving a different response helps you learn the layout more
quickly.

For users that need to operate in sound-sensitive environments like
open offices or libraries,
[[http://deskthority.net/wiki/Matias_switch#Quiet_click][Matias Quiet
Click]] switches are a popular choice since they still offer tactility
without the noise. Other users prefer switches in the
[[http://deskthority.net/wiki/Cherry_MX][Cherry MX]] family, which use
different keycaps and switch plates but still work fine.

*** Diodes

In order to avoid ghosting, each switch needs a diode. The [[https://www.radioshack.com/product/index.jsp?productId=2062587][1N4148]] is a
readily-available choice, but nearly any signal diode would work.

*** Microcontroller

The circuit board design uses a [[http://www.pololu.com/product/3101][Pololu A-star micro]]. Hand-wired boards
can also use a [[http://www.pjrc.com/teensy/index.html][Teensy 2]] or [[http://arduino.cc/en/Main/ArduinoBoardMicro][Arduino Pro Micro]].

Be sure to get a microcontroller without headers so it will fit in
between the bottom layer and the plate. USB micro is preferred over
USB mini for this reason as well.

*** Keycaps

Caps for Matias switches are included in the official kits. Sculpted
caps are also available [[http://matias.ca/order/#keycaps][from Matias]] or by harvesting from old Alps
keyboards. It's recommended that you use unlabeled keys, because due
to the different sizes and orientations of certain keys (backspace,
shift, enter, etc) many of the labels will be incorrect if present.

Cherry switches have more options. This
[[http://pimpmykeyboard.com/dsa-pbt-abs-blank-keycap-sets/][DSA-shaped base set]] (spherical indentations on the key, same profile for each
row) from Signature Plastics has 52 1x keys plus a few extras we won't
use. There are two "deep dish" keys in that set which you can place
under your index fingers on the home row to help guide your hands to
the right spot without looking. However, you only get a single 1.5x
keycap, and the middle two thumb keys both use them, so you might want
to pick up an extra.

** Circuit Board

The =atreus.rkt= program calculates switch and diode positions based
on row/column counts, spacing, and rotation factors, and emits a
=atreus.kicad_pcb= file. The board outline and traces are done by hand
and are stored in the =header.rktd= and =traces.rktd= files
respectively. The =atreus.kicad_pcb= file can be imported into [[http://kicad-pcb.org][Kicad]]
which can export Gerber files suitable for fabrication. A copy of the
Kicad PCB file is included in the repository if you don't want to
recompile it using Racket. Recompiling is only required if you want to
make changes to the procedurally-generated portions of the board.

Unfortunately most PCB fabricators require a minimum order of 10 or
so, making this impractical for one-offs. The PCB is not required, so
for one-off boards it's usually more sensible to stick with a [[http://atreus.technomancy.us/assembly-hand-wired.pdf][hand-wired build]] instead.

** Case

Layered laser-cut wood or acrylic. The [[http://geekhack.org/index.php?topic%3D54759.msg1304117#msg1304117][mark II case]] (EPS files in the
=case/= directory) features 8 screw holes and a kind of "stair step"
design around the top and bottom of the key clusters; mark I has
straight lines.

The files =alps-top-plate-3mm.eps=, =alps-switch-plate-3mm.eps=, and
=bottom-plate-3mm.eps= are to be cut in 3mm acrylic or wood. The
=spacer.eps= file should be cut on something thicker; between 4.5mm
and 6mm is recommended. The spacer needs to be at least as thick as
the connector of the USB cable you're using. If you are using Cherry
switches, you should use =cherry-3mm.eps= for the top, switch, and
bottom plates instead, but the spacer is the same.

There is also a programmatically-implemented version of the case
written in OpenSCAD; it is more flexible (you can tweak the number of
rows/cols, etc and recompile) but it doesn't match the canonical case
exactly; in particular the screw holes are placed differently.

The original case (=case-mk-i.svg=) design is also included; it is
slightly less wide and has a minor asymmetry with the screws on the
bottom side.

On a 100W Epilog laser, the 3mm layers cut in about a minute and a
half. I did a run with 6mm acrylic of the other layers which took
nearly 6 minutes.

Wood cases should be finished with sandpaper and finishing oil/wax or
with [[https://atreus.technomancy.us/lacquer.pdf][lacquer]] which
takes longer but feels nicer.

** Firmware

The [[https://atreus.technomancy.us/tmk][TMK firmware]] is
recommended. You should be able to change into the =keyboard/atreus=
directory and run =make KEYMAP=qwerty= (or whichever variant you want)
to produce a qwerty =atreus.hex= file. You will probably want to
create your own layout once you've gotten a chance to try it and see
what works for you. See the readme for instructions of how to upload
it to the keyboard's controller.

There is also the older [[https://github.com/technomancy/atreus-firmware][atreus-firmware]] custom codebase which works
but has fewer features. It is recommended mostly for learning purposes
since the code is much simpler and easier to understand than TMK.

** Bill of Materials

If you don't want to get [[https://atreus.technomancy.us][the kit]] you can source parts yourself. (You
can also order a partial kit if you want to customize some of the
parts.) Using Cherry switches is cheapest unless you can find cheap
Alps-mount keycaps:

- 50 Matias or Cherry switches from mechanicalkeyboards.com
- 50 diodes from radio shack, ebay, or whatever
- 40 1x, 2 1.5x Keycaps from signatureplastics.com OR full keycap set from matias.ca
- A-star micro from pololu.com
- Case materials from various sources
- Case laser cutting: 7.5 minutes on a 100W Epilog laser; varies by source
- USB micro cable, get anywhere

Recommended but optional:

- 5 linear Matias or Cherry switches

The base keycap set from Signature Plastics only has one 1.5x key,
which is used for the inner thumb keys. You can use a 1x key for one
of them, but it looks kind of tacky, so I recommend getting a second
1.5x keycap separately.

*** Other Tools

You'll need a soldering iron, solder, and a wire cutter. A multimeter
can come in handy for testing the connections but is optional. You'll
also need eight M3 machine screws with nuts; the length of the screws
depends on the thickness of the acrylic you use. You can add rubber
feet to the bottom to prevent the board from sliding around when
placed on a desk. You'll also need sandpaper and lacquer for the
wooden case.

If you are building a hand-wired board you will also need a glue gun,
hookup wire, and wire strippers.

** Assembly

See the [[https://atreus.technomancy.us/assembly.pdf][assembly instructions]] PDF.

Hand-wired boards will want the [[https://atreus.technomancy.us/assembly-hand-wired.pdf][previous edition of the assembly instructions]].

The LaTeX source to the assembly instructions is in the =assembly= directory.

** Inspiration

These fine projects all provided inspiration for various aspects of
the Atreus, as well as the folks on the =#geekhack= freenode channel.

- [[https://geekhack.org/index.php?topic=22780.0][Ergodox]]
- [[http://deskthority.net/workshop-f7/onehand-20-keyboard-t6617.html][OneHand]]
- [[http://blog.fsck.com/2013/12/better-and-better-keyboards.html][keyboard.io]]
- [[http://geekhack.org/index.php?topic=48718][ErgoT]]

** Builds

If you've built an Atreus, please add your name to [[https://github.com/technomancy/atreus/wiki/BuildLogs][the build logs wiki]].

There's also a [[https://atreus.technomancy.us/list][mailing list]] for people who have built or ordered an
Atreus or are interested in doing so.

** Orestes

A new [[https://www.flickr.com/photos/technomancy/14654421878][experimental build]] uses the [[http://pjrc.com/store/teensy31.html][Teensy 3]] microcontroller and
ARM [[https://github.com/technomancy/orestes/tree/teensy3][Forth-based]] firmware, but this is not yet suitable for general-purpose use.

** License

Copyright © 2014-2016 Phil Hagelberg and contributors

Released under the [[https://www.gnu.org/licenses/gpl.html][GNU GPL version 3]]


changelog.md
# Change log

This is a listing of all changes and prototypes for the Atreus
keyboard design. Changes to the
[Atreus Deck](https://atreus.technomancy.us/marki) are elsewhere.

## Keyboardio Atreus: 2020-03-17

* Completely redesigned factory-produced case.
* Metal switch plate.
* USB C port.
* Optionally labeled XDA keycaps.
* 15mm smaller in two dimensions.

## NX-04: 2019-05-25

This was the first prototype to incorporate the redesigned PCB by
Jesse Vincent.

* 2 additional keys.
* wooden case with 4 layers instead of 5.
* detachable USB cable.
* swappable Cherry MX switches.
* surface-mount controller and diodes.
* Kaleidoscope is the default firmware.

## Mark 3.8: 2018-7-28

* QMK is the default firmware for all boards now.

## Mark 3.7: 2017-6-29

* Manually bending middle switch pins on Alps is no longer necessary.
* Spacer layer is split into three pieces to be less wasteful with materials.

## Mark 3.6: 2016-3-14

* TMK is the default firmware.
* New layout with arrows on the fn layer.
* PCB is now label-side up by default.

## Mark 3.5: 2015-6-23

* Change to Matias switches.

## NX-03: 2015-05-08

* Matias Clicky switches.
* Alps-compatible switch plate.

## Mark 3.1: 2014-11-25

* [Second revision of circuit board](http://gerblook.org/pcb/jweM8BcTbgKCL4zsj8tEYj).
* Arrows on L2 shifted over one column.
* Circuit board license changed to match the rest of the Atreus project.

The second revision of the circuit board exposes the reset pin,
removing the need for hookup wire described below. It also removes
[the need to cut a notch in the circuit board](https://www.flickr.com/photos/technomancy/15862732742)
to allow the USB connector more clearance.

The layout changed slightly by shifting the arrow keys on L2 over one
column so they can be hit from the normal position without the pinky key.

## Mark 3: 2014-10-30

* Initial [board design](http://gerblook.org/pcb/zqMY66AEsKz3xehSrdoDV6) in KiCAD + Racket.
* Assembly guide updated to include wood finishing.
* Cable hole in case spacer has rounded edges.
* New layout: multidvorak.
* Switch to Pololu A-Star Micro controller. (also USB mini->micro)

A
[circuit board was designed in KiCAD with the switch placement logic written in Racket](http://technomancy.us/176). The
first revision of the circuit board does not expose the reset pin, so
hookup wire is needed to provide a backup reset mechanism once the board
is fully assembled. The firmware contains a reset mechanism, but if
bugs in the firmware prevent it from being activated, the back case
can be removed and the hookup wires can be connected twice in a second
to initiate a reset.

The new design replaced the Teensy 2 microcontroller with the smaller
[A-Star Micro](http://www.pololu.com/product/3101) which uses a USB
Micro connector instead of USB Mini.

The new multidvorak layout allows the firmware to support hardware
dvorak and software dvorak simultaneously. The edges of the cable hole
are rounded off to reduce strain on the cable.

## NX-02: 2014-08-05

* Matias Quiet Click switches.
* Teensy 3: ARM-based microcontroller.
* Orestes: firmware written in Forth.

An
[experimental prototype](https://www.flickr.com/photos/technomancy/14654421878)
using Matias Quiet Click switches and the ARM-based Teensy 3
controller. The Matias switch shape was supported by cutting the
notches out of each switch hole individually with a knife.

Porting the firmware to the ARM platform was accompanied by porting
[an implementation of Forth](https://github.com/technomancy/orestes/tree/teensy3)
and porting the firmware logic to that.

## Mark 2: 2014-04-22

* New OpenSCAD [case design](https://www.flickr.com/photos/technomancy/14719013574) by Constantine.
* New layouts: hardware dvorak, hardware colemak.
* Assembly guide.

The case was
[redesigned in OpenSCAD](https://github.com/technomancy/atreus/blob/master/case/openscad/atreus_case.scad)
to be easier to produce from mail-order laser cutting services. This
revision can be recognized by its use of eight screws instead of the
original seven. The cuts in the top plate around the switches are also
much closer to the switches themselves.
[A second independent build](http://geekhack.org/index.php?topic=57007.msg1292144)
was documented.

The assembly process is documented properly.

## Mark 1: 2014-04-01

* Custom Atreus firmware in C.
* Layout changes; new layer.

A
[custom Atreus-specific firmware codebase](https://github.com/technomancy/atreus-firmware)
was started. Digits were moved to a numpad arrangement and the L2
layer (function and arrow keys) was added. The first
[third-party build](http://blog.tarn-vedra.de/2014/04/reproducing-keyboard-science_1707.html)
was documented.

## NX-01: 2014-02-22

* [Initial prototype](https://www.flickr.com/photos/technomancy/12805226463)!
* Case hand-drawn in Inkscape.
* Teensy 2 microcontroller, hand-wired matrix.
* Uses TMK firmware.

This was the first Atreus,
[created solely for personal use](http://technomancy.us/173) and
meticulously hand-wired with a number of errors. Support for the
Atreus was
[added to the TMK firmware codebase](https://github.com/technomancy/tmk_keyboard/tree/atreus),
but power-saving bugs in TMK meant that it occasionally dropped key
presses on low-voltage laptops.


case/case.rkt
#lang racket
;; Atreus 2 case design
;; Copyright © 2019 Phil Hagelberg and contributors
;; released under the GPLv3 or later

;; TODO:
;; * port notch on bottom plate and spacer
;; * top layer

(require xml)

;; glowforge uses 96 dpi, 25.4 mm in an inch
(define scale (/ 96 25.4))
(define width 260)
(define height 132)


(define cols 6) ; per hand
(define rows 4)
(define angle (degrees->radians 10))
(define corner-radius 6.0)

(define alps-switch-width 15.887)
(define alps-switch-height 13.087)
(define cherry-switch-width 13.62)
(define cherry-switch-height 13.72)
(define cherry? false)
(define switch-height (if cherry? cherry-switch-height alps-switch-height))
(define switch-width (if cherry? cherry-switch-width alps-switch-width))

(define switch-spacing 19.0)

(define screw-radius 1.4) ; for M3 screws + kerf

(define side-screw-distance (* switch-spacing rows))
(define bottom-screw-distance (* switch-spacing cols))
(define left corner-radius)
(define bottom 95) ; outer bottom
(define left-top (+ left (* side-screw-distance (sin angle))))
(define top (- bottom (* side-screw-distance (cos angle))))
(define right (- width corner-radius))
(define right-top (- right (* side-screw-distance (sin angle))))
(define mid-bottom (+ bottom (* bottom-screw-distance (sin angle)) -3))
(define mid-offset 25)
(define mid-x (/ width 2))
(define mid-left (- mid-x mid-offset))
(define mid-right (+ mid-x mid-offset))

(define hull-coords (list (list right-top top)
                          (list right bottom)
                          (list mid-right mid-bottom)
                          (list mid-left mid-bottom)
                          (list left bottom)
                          (list left-top top)))

;;; screws
(define screws
  `(g () ,@(for/list ([s (append (take hull-coords 2)
                                 ;; the bottom middle has only one screw but
                                 ;; two hull positions
                                 (list (list (/ width 2) mid-bottom))
                                 (drop hull-coords 4))])
             `(circle ((r ,(number->string screw-radius))
                       (cx ,(number->string (first s)))
                       (cy ,(number->string (second s))))))))

;;; outline
(define outline-coords (append hull-coords (take hull-coords 2)))

(define (to-next-screw? theta current-screw)
  (let* ([current (list-ref outline-coords current-screw)]
         [cx (first current)] [cy (second current)]
         [next (list-ref outline-coords (add1 current-screw))]
         [nx (first next)] [ny (second next)]
         [dx (- nx cx)] [dy (- ny cy)]
         [next-theta (- (radians->degrees (atan dy dx)))])
    (= (floor (modulo (floor next-theta) 180))
       (floor (modulo (- theta 90) 180)))))

;; trace the outline by going from screw to screw until you've gone full-circle
(define (outline-points coords theta current-screw)
  (if (< -360 (- theta 90) 360)
      (let* ([current (list-ref outline-coords current-screw)]
             [sx (first current)] [sy (second current)]
             [x (+ sx (* (cos (degrees->radians theta)) corner-radius))]
             [y (- sy (* (sin (degrees->radians theta)) corner-radius))]
             [coords (cons (format "~s,~s" x y) coords)])
        (if (to-next-screw? theta current-screw)
            (outline-points coords theta (add1 current-screw))
            (outline-points coords (sub1 theta) current-screw)))
      coords))

(define port-depth 8)

(define port-curve (list (format "~s,~s" mid-right (- top corner-radius))
                         (format "~s,~s" (+ mid-x corner-radius)
                                 (+ top port-depth
                                    (- corner-radius)))
                         (format "~s,~s" (- mid-x corner-radius)
                                 (+ top port-depth
                                    (- corner-radius)))
                         (format "~s,~s" mid-left (- top corner-radius))))

(define (outline with-port?)
  `(polygon ((points ,(string-join (let ((noport (outline-points '() 90 0)))
                                     (if with-port?
                                         (append noport port-curve)
                                         noport)))))))

;;; switches

(define column-offsets `(8 5 0 6 11 ,(+ 8 switch-spacing switch-spacing)))

(define (switch row col)
  (let* ([x (* (+ 1 col) switch-spacing)]
         [y (+ (list-ref column-offsets col) (* switch-spacing row))])
    `(rect ((height ,(number->string switch-height))
            (width ,(number->string switch-width))
            (x ,(number->string x))
            (y ,(number->string y))))))

(define hand-height (+ (* switch-spacing rows) (- switch-spacing switch-height)
                       (list-ref column-offsets 0)))
(define switch-x-offset -6.5)
(define switch-y-offset (- bottom hand-height -3.5))

(define switches
  `(g ((transform ,(format "translate(~s, ~s) rotate(~s, ~s, ~s)"
                           switch-x-offset switch-y-offset
                           (radians->degrees angle)
                           0 hand-height)))
      ,@(for/list ([col (in-range cols)]
                   #:when true
                   [row (if (= 5 col) '(0 1) (in-range rows))])
          (switch row col))))

(define switches-right
  `(g ((transform ,(format "translate(~s,~s) scale(-1, 1)" width 0)))
      ,switches))

(define logo-doc (call-with-input-file "logo-fragment.svg" read-xml))

(define pcb-doc (call-with-input-file "pcb-fragment.svg" read-xml))

(define (layer plate)
  (document (prolog '() false '())
            (xexpr->xml
             `(svg ((xmlns:svg "http://www.w3.org/2000/svg")
                    (height ,(number->string (* height scale)))
                    (width ,(number->string (* width scale))))
                   ,@(if (eq? plate 'switch)
                         `((g ((transform "translate(436, 115)")
                               (stroke "red"))
                              ,(xml->xexpr (document-element logo-doc))))
                         '())
                   ,@(if (eq? plate 'spacer)
                         (list (xml->xexpr (document-element pcb-doc)))
                         (list))
                   (g ((transform ,(format "scale(~s, ~s)" scale scale))
                       (stroke-width "1")
                       (stroke "black")
                       (fill-opacity "0"))
                      ,screws
                      ,(outline (not (eq? plate 'switch)))
                      ,@(if (eq? plate 'switch)
                            (list switches switches-right)
                            (list)))))
            '()))

;; to laser cut these, you have to open them in inkscape, then save,
;; and then upload; for some reason the glowforge crashes if you try to cut
;; them directly.

(define (write-out-layer layer-name)
  (call-with-output-file (format "case2-~a.svg" (symbol->string layer-name))
    (lambda (out)
      (display "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>" out)
      (display-xml (layer layer-name) out))
    #:exists 'replace))

;; live-reload with:
;; qiv --watch case2-switch.svg

(write-out-layer 'switch)
(write-out-layer 'bottom)
(write-out-layer 'spacer)


atreus.rkt
#lang racket

(define cols 12)
(define rows 4)

(define x-offset 20)
(define y-offset 20)

(define spacing 19)
(define angle 10)

(define column-offsets '(8 5 0 6 11 53 53 11 6 0 5 8))

(define (switch-module x y rotation label net-pos net-neg)
  ;; TODO: set timestamps?
  `(module MX_FLIP (layer Front) (tedit 4FD81CDD) (tstamp 543EF801)
    (at ,x ,y ,rotation)
    (path /543DB910) ; TODO: this is not documented; no idea what it does
    (fp_text reference ,label (at 0 3.302 ,rotation) (layer F.SilkS)
             (effects (font (size 1.524 1.778) (thickness 0.254))))
    (fp_line (start -6.35 -6.35) (end 6.35 -6.35)
             (layer F.SilkS) (width 0.381))
    (fp_line (start 6.35 -6.35) (end 6.35 6.35)
             (layer F.SilkS) (width 0.381))
    (fp_line (start 6.35 6.35) (end -6.35 6.35)
             (layer F.SilkS) (width 0.381))
    (fp_line (start -6.35 6.35) (end -6.35 -6.35)
             (layer F.SilkS) (width 0.381))
    (pad 0 np_thru_hole circle (at 0 0) (size 3.9878 3.9878)
         (drill 3.9878)) ; switch hole, no copper
    (pad 0 np_thru_hole circle (at -5.08 0) (size 1.7018 1.7018)
         (drill 1.7018)) ; board-mount hole, no copper
    (pad 0 np_thru_hole circle (at 5.08 0) (size 1.7018 1.7018)
         (drill 1.7018)) ; board-mount hole, no copper
    (pad 1 thru_hole circle (at 2.54 -5.08) (size 2.286 2.286) (drill 1.4986)
         (layers *.Cu *.SilkS *.Mask) ,net-pos)
    (pad 1 thru_hole circle (at 3.81 -2.54) (size 2.286 2.286) (drill 1.4986)
         (layers *.Cu *.SilkS *.Mask) ,net-pos)
    (pad 2 thru_hole circle (at -2.54 -5.08) (size 2.286 2.286) (drill 1.4986)
         (layers *.Cu *.SilkS *.Mask) ,net-neg)
    (pad 2 thru_hole circle (at -3.81 -2.54) (size 2.286 2.286) (drill 1.4986)
         (layers *.Cu *.SilkS *.Mask) ,net-neg)))

(define (diode-module x y rotation label net-pos net-neg)
  `(module DIODE (layer Front) (tedit 4E0F7A99) (tstamp 543EF854)
    (at ,x ,y ,(+ 90 rotation))
    (path /543DB90F)
    (fp_text reference D2:2 (at 0 0 180) (layer F.SilkS) hide
             (effects (font (size 1.016 1.016) (thickness 0.2032))))
    (fp_line (start -1.524 -1.143) (end 1.524 -1.143)
             (layer F.SilkS) (width 0.2032))
    (fp_line (start 1.524 -1.143) (end 1.524 1.143)
             (layer F.SilkS) (width 0.2032))
    (fp_line (start 0 -1.143) (end 0 1.143)
             (layer F.SilkS) (width 0.2032))
    (fp_line (start 0 -1.143) (end -1.524 0)
             (layer F.SilkS) (width 0.2032))
    (fp_line (start -1.524 0) (end 0 1.143)
             (layer F.SilkS) (width 0.2032))
    (fp_line (start 1.524 1.143) (end -1.524 1.143)
             (layer F.SilkS) (width 0.2032))
    (fp_line (start -1.524 1.143) (end -1.524 -1.143)
             (layer F.SilkS) (width 0.2032))
    (fp_line (start -3.81 0) (end -1.6637 0) (layer Back) (width 0.6096))
    (fp_line (start 1.6637 0) (end 3.81 0) (layer Back) (width 0.6096))
    (fp_line (start -3.81 0) (end -1.6637 0) (layer Front) (width 0.6096))
    (fp_line (start 1.6637 0) (end 3.81 0) (layer Front) (width 0.6096))

    (pad 1 thru_hole circle (at -3.81 0 180) (size 1.651 1.651)
         (drill 0.9906) (layers *.Cu *.SilkS *.Mask) ,net-neg)
    (pad 2 thru_hole rect (at 3.81 0 ,rotation) (size 1.651 1.651)
         (drill 0.9906) (layers *.Cu *.SilkS *.Mask) ,net-pos)
    (pad 99 smd rect (at -1.6637 0 ,rotation) (size 0.8382 0.8382)
         (layers Front F.Paste F.Mask))
    (pad 99 smd rect (at -1.6637 0 ,rotation) (size 0.8382 0.8382)
         (layers Back B.Paste B.Mask))
    (pad 99 smd rect (at 1.6637 0 ,rotation) (size 0.8382 0.8382)
         (layers Front F.Paste F.Mask))
    (pad 99 smd rect (at 1.6637 0 ,rotation) (size 0.8382 0.8382)
         (layers Back B.Paste B.Mask))))

(define microcontroller-module
  `(module A_STAR (layer Front) (tedit 4FDC31C8) (tstamp 543EF800)
    (at 134 65 270)
    (path /543EEB02)
    (fp_line (start 10.1 7.62) (end 10.1 -7.62) (layer F.SilkS) (width 0.381))
    (fp_line (start 10.1 -7.62) (end -15.24 -7.62) (layer F.SilkS) (width 0.381))

    (pad B5 thru_hole circle (at -13.97 6.35 270) (size 1.7526 1.7526)
         (drill 1.0922) (layers *.Cu *.SilkS *.Mask) (net 7 N-col-2))
    (pad B4 thru_hole circle (at -11.43 6.35 270) (size 1.7526 1.7526)
         (drill 1.0922) (layers *.Cu *.SilkS *.Mask) (net 8 N-col-3))
    (pad E6 thru_hole circle (at -8.89 6.35 270) (size 1.7526 1.7526)
         (drill 1.0922) (layers *.Cu *.SilkS *.Mask) (net 9 N-col-4))
    (pad D7 thru_hole circle (at -6.35 6.35 270) (size 1.7526 1.7526)
         (drill 1.0922) (layers *.Cu *.SilkS *.Mask) (net 5 N-col-0))
    (pad C6 thru_hole circle (at -3.81 6.35 270) (size 1.7526 1.7526)
         (drill 1.0922) (layers *.Cu *.SilkS *.Mask) (net 6 N-col-1))
    (pad D4 thru_hole circle (at -1.27 6.35 270) (size 1.7526 1.7526)
         (drill 1.0922) (layers *.Cu *.SilkS *.Mask) (net 10 N-col-5))

    (pad D0 thru_hole circle (at 1.27 6.35 270) (size 1.7526 1.7526)
         (drill 1.0922) (layers *.Cu *.SilkS *.Mask) (net 1 N-row-0))
    (pad D1 thru_hole circle (at 3.81 6.35 270) (size 1.7526 1.7526)
         (drill 1.0922) (layers *.Cu *.SilkS *.Mask) (net 2 N-row-1))
    (pad D3 thru_hole circle (at 6.35 6.35 270) (size 1.7526 1.7526)
         (drill 1.0922) (layers *.Cu *.SilkS *.Mask) (net 3 N-row-2))
    (pad D2 thru_hole circle (at 8.89 6.35 270) (size 1.7526 1.7526)
         (drill 1.0922) (layers *.Cu *.SilkS *.Mask) (net 4 N-row-3))

    (pad F7 thru_hole circle (at -13.97 -6.35 270) (size 1.7526 1.7526)
         (drill 1.0922) (layers *.Cu *.SilkS *.Mask) (net 13 N-col-8))
    (pad F6 thru_hole circle (at -11.43 -6.35 270) (size 1.7526 1.7526)
         (drill 1.0922) (layers *.Cu *.SilkS *.Mask) (net 12 N-col-7))
    (pad B6 thru_hole circle (at -8.89 -6.35 270) (size 1.7526 1.7526)
         (drill 1.0922) (layers *.Cu *.SilkS *.Mask) (net 11 N-col-6))
    (pad B7 thru_hole circle (at -6.35 -6.35 270) (size 1.7526 1.7526)
         (drill 1.0922) (layers *.Cu *.SilkS *.Mask) (net 15 N-col-10))
    (pad D6 thru_hole circle (at -3.81 -6.35 270) (size 1.7526 1.7526)
         (drill 1.0922) (layers *.Cu *.SilkS *.Mask) (net 14 N-col-9))
    (pad RST thru_hole circle (at -1.27 -6.35 270) (size 1.7526 1.7526)
       (drill 1.0922) (layers *.Cu *.SilkS *.Mask) (net 0 ""))
    (pad GND thru_hole circle (at 6.35 -6.35 270) (size 1.7526 1.7526)
         (drill 1.0922) (layers *.Cu *.SilkS *.Mask) (net 0 ""))))

(define nets
  `((net 0 "")
    (net 1 N-row-0)
    (net 2 N-row-1)
    (net 3 N-row-2)
    (net 4 N-row-3)
    (net 5  N-col-0)
    (net 6  N-col-1)
    (net 7  N-col-2)
    (net 8  N-col-3)
    (net 9  N-col-4)
    (net 10 N-col-5)
    (net 11 N-col-6)
    (net 12 N-col-7)
    (net 13 N-col-8)
    (net 14 N-col-9)
    (net 15 N-col-10)
    ,@(for/list ([s (in-range 42)])
        (list 'net (+ 16 s) (string->symbol (format "N-diode-~s" s))))))

(define (net-class nets)
  (append '(net_class Default "This is the default net class."
            (clearance 0.254)
            (trace_width 0.2032)
            (via_dia 0.889)
            (via_drill 0.635)
            (uvia_dia 0.508)
            (uvia_drill 0.127))
          (for/list ([n nets])
            (list 'add_net (last n)))))

(define (switch row col)
  (let* ([left? (< col 6)]
         [rotation (if left? -10 10)]
         [x (* (+ 1 col) spacing)]
         [y (+ (list-ref column-offsets col) (* spacing row))]
         [hypotenuse (sqrt (+ (* x x) (* y y)))]
         [Θ (atan (/ y x))]
         [Θ′ (- Θ (degrees->radians rotation))]
         [x′ (+ (if left? x-offset 5) (* hypotenuse (cos Θ′)))]
         [y′ (+ (if left? y-offset (+ y-offset 42.885)) (* hypotenuse (sin Θ′)))]
         [label (format "SW~a:~a" col row)]
         [diode (+ row (* col 4))]
         ;; if we try to number nets linearly, kicad segfaults; woo!
         ;; so we re-use the nets we skipped with the missing col 5/6 diodes
         [diode (cond [(> diode 44) (- diode 20)]
                      [(> diode 41) (- diode 21)]
                      [true diode])]
         [net-col (if left? col (- col 1))]
         [diode-net `(net ,(+ 16 diode)
                      ,(string->symbol (format "N-diode-~s" diode)))]
         [column-net `(net ,(+ net-col 5)
                       ,(string->symbol (format "N-col-~s" net-col)))])
    (switch-module x′ y′ rotation label
                   (if left? diode-net column-net)
                   (if left? column-net diode-net))))

(define (diode row col)
  (let* ([left? (< col 6)]
         [rotation (if left? -10 10)]
         [x (* (+ 1 col) spacing)]
         [y (+ (list-ref column-offsets col) (* spacing row))]
         [hypotenuse (sqrt (+ (* x x) (* y y)))]
         [Θ (atan (/ y x))]
         [Θ′ (- Θ (degrees->radians rotation))]
         [x′ (+ (if left? x-offset 5) (* hypotenuse (cos Θ′))
                (if left? 9 -9))]
         [y′ (+ (if left? y-offset (+ y-offset 42.885))
                (* hypotenuse (sin Θ′)))]
         [label (format "D~a:~a" col row)]
         [diode (+ row (* col 4))]
         ;; if we try to number nets linearly, kicad segfaults; woo!
         ;; so we re-use the nets we skipped with the missing col 5/6 diodes
         [diode (cond [(> diode 44) (- diode 20)]
                      [(> diode 41) (- diode 21)]
                      [true diode])]
         [net-row (cond [(= col 5) 2]
                        [(= col 6) 3]
                        [true row])])
    (diode-module x′ y′ rotation label
                  `(net ,(+ 16 diode)
                    ,(string->symbol (format "N-diode-~s" diode)))
                  `(net ,(+ net-row 1)
                    ,(string->symbol (format "N-row-~s" net-row))))))

(define switches+diodes
  (for/list ([col (in-range cols)]
             #:when true
             [row (if (or (= 5 col) (= 6 col))
                      '(0 1) (in-range rows))])
    (list (switch row col) (diode row col))))

(define edge-cuts
  (for/list [(s '([31 22] [84 22] [127 30]
                          [129 49] [129 75] [139 75]
                  [139 49] [141 30] [185 22]
                  [237 22] [250 95] [161 112] [107 112] [18 95]))
             (e '([84 22] [127 30]
                          [129 49] [129 75] [139 75] [139 49]
                  [141 30] [185 22]
                  [237 22] [250 95] [161 112] [107 112] [18 95] [31 22]))]
    `(gr_line (start ,@s) (end ,@e) (angle 90) (layer Edge.Cuts) (width 0.3))))

(define board
  (apply append nets
         (list (net-class nets))
         (list microcontroller-module)
         edge-cuts
         switches+diodes))

(define (write-placement filename)
  (when (file-exists? filename) (delete-file filename))
  (call-with-output-file filename
    (λ (op)
      (display (call-with-input-file "header.rktd"
                 (curry read-string 9999)) op)
      ;; kicad has this terrible bug where it's whitespace-sensitive here =(
      (display "\n" op)
      (for ([f board])
        (pretty-print f op 1))
      (display (call-with-input-file "traces.rktd"
                 (curry read-string 999999)) op)
      (display ")" op))))

;; TODOs:

;; controller traces
;; middle switch traces
;; middle switch diodes below

(write-placement "atreus.kicad_pcb")
